### PR TITLE
Bluetooth: GATT: Reduce footprint of client parameters

### PR DESCRIPTION
**Pull request by Vudentz**
This reduces the overall footprint of GATT client parameters so they don't carry bt_att_req themselves, it also optimize bt_gatt_subscribe_params so they don't have to carry the peer address themselves thus reducing their footprint even further.

### CODE REVIEW

File: subsys/bluetooth/host/att.c [hunk 1]
```diff
@@ -61,6 +61,9 @@ NET_BUF_POOL_DEFINE(prep_pool, CONFIG_BT_ATT_PREPARE_COUNT, BT_ATT_MTU,
 		    sizeof(struct bt_attr_data), NULL);
 #endif /* CONFIG_BT_ATT_PREPARE_COUNT */
 
+K_MEM_POOL_DEFINE(req_pool, sizeof(struct bt_att_req),
+		  sizeof(struct bt_att_req), CONFIG_BT_ATT_TX_MAX, 16);
```
Comment (jhedberg) [inline subsys/bluetooth/host/att.c]: Wouldn't `k_mem_slab` be the more appropriate data structure to use since you're allocating fixed same-sized blocks of data? Using `k_mem_pool` has the added cost of supporting arbitrary-sized allocations which you're not taking advantage of.

**Review by jhedberg — CHANGES_REQUESTED** [review]

File: subsys/bluetooth/host/att.c (hunk 1 above)
Comment (Vudentz) [reply subsys/bluetooth/host/att.c]: Will change to k_mem_slab.

Comment (Vudentz) [reply subsys/bluetooth/host/att.c]: Done.

File: subsys/bluetooth/host/att_internal.h [hunk 2]
```diff
@@ -6,6 +6,7 @@
  * SPDX-License-Identifier: Apache-2.0
  */
 
+#define BT_EATT_PSM		0x7f
```
Comment (jhedberg) [inline subsys/bluetooth/host/att_internal.h]: Seems unrelated to this PR?

File: subsys/bluetooth/host/att_internal.h [hunk 3]
```diff
@@ -222,6 +224,26 @@ struct bt_att_signature {
...
+} __packed;
+
+/* Handle Multiple Value Notification */
+#define BT_ATT_OP_NOTIFY_MULT			0x23
+struct bt_att_notify_mult {
+	u16_t handle;
+	u16_t len;
+	u8_t  value[0];
+} __packed;
+
```
Comment (jhedberg) [inline subsys/bluetooth/host/att_internal.h]: These also seem unrelated?

File: subsys/bluetooth/host/att_internal.h [hunk 4]
```diff
@@ -139,6 +140,7 @@ struct bt_att_read_mult_req {
 /* Read Multiple Respose */
 #define BT_ATT_OP_READ_MULT_RSP			0x0f
 struct bt_att_read_mult_rsp {
+	u16_t len;
```
Comment (jhedberg) [inline subsys/bluetooth/host/att_internal.h]: Unrelated?

**Review by jhedberg — CHANGES_REQUESTED** [review]
These seems to be a bunch of unrelated changes that have slipped into this PR from your EATT work.

File: subsys/bluetooth/host/att_internal.h (hunk 4 above)
Comment (Vudentz) [reply subsys/bluetooth/host/att_internal.h]: Fixed.

File: subsys/bluetooth/host/att_internal.h (hunk 3 above)
Comment (Vudentz) [reply subsys/bluetooth/host/att_internal.h]: Fixed.

File: subsys/bluetooth/host/gatt.c [hunk 5]
```diff
@@ -2171,16 +2186,58 @@ bool bt_gatt_is_subscribed(struct bt_conn *conn,
...
+	sub = gatt_sub_find(conn);
+	if (sub) {
+		return sub;
+	}
+
+	sub = gatt_sub_find(NULL);
+	if (sub) {
+		bt_addr_le_copy(&sub->peer, &conn->le.dst);
+		return sub;
+	}
```
Comment (jhedberg) [inline subsys/bluetooth/host/gatt.c]: The pattern I've used elsewhere for this, in order to avoid iterating the list twice, is to pass a second "free_slot" parameter to the lookup function. The lookup function will set that to point to the first empty slot in case the lookup of the actual connection fails: https://github.com/zephyrproject-rtos/zephyr/blob/8067aca7e90617fb49fb0895066c164d60559198/subsys/bluetooth/mesh/cfg_srv.c#L1107-L1131

Would something like that make sense here?

Comment (Vudentz) [reply subsys/bluetooth/host/gatt.c]: To be honest I haven't consider this would be a big deal since MAX_CONN+MAX_PAIRED is normally not that big, but I can use the same logic to just return an unused slot in one go.

Comment (Vudentz) [reply subsys/bluetooth/host/gatt.c]: Btw an alternative for this is to use ffs and then maintain a bitmap i.e:

```
/* Find unique id in range from 1 to max but no bigger then
 * sizeof(int) * 8. ffs() is used since it is POSIX standard
 */
uint8_t util_get_uid(unsigned int *bitmap, uint8_t max)
{
	uint8_t id;

	id = ffs(~*bitmap);

	if (!id || id > max)
		return 0;

	*bitmap |= 1 << (id - 1);

	return id;
}

/* Clear id bit in bitmap */
void util_clear_uid(unsigned int *bitmap, uint8_t id)
{
	if (!id)
		return;

	*bitmap &= ~(1 << (id - 1));
}
```

**Comment by Vudentz** [comment]
@jhedberg Anything else that needs fixing before we merge this one?

**Review by wdhpawa — APPROVED** [review]
Looks OK to me.

**Comment by joerchan** [comment]
@Vudentz Did you want to do the ffs bitmap thing? Because I think we could CLZ/CTZ as well for that (if  I understood correctly). The are implemented as:  u32_count_leading_zeros and u32_count_trailing_zeros.

**Comment by jhedberg** [comment]
> @jhedberg Anything else that needs fixing before we merge this one?

@Vudentz I think @joerchan pointed out some valid issues, so I'd wait for their resolution first

File: subsys/bluetooth/host/att.c [hunk 6]
```diff
@@ -85,7 +88,7 @@ struct bt_att {
 #endif
 };
 
-static struct bt_att bt_req_pool[CONFIG_BT_MAX_CONN];
+static struct bt_att bt_req_slab[CONFIG_BT_MAX_CONN];
```
[the remark this answers is not in the document]
Comment (Vudentz) [reply subsys/bluetooth/host/att.c]: Im fixing it was unintented change.

File: subsys/bluetooth/host/gatt.c [hunk 7]
```diff
@@ -1602,13 +1607,23 @@ static int gatt_send(struct bt_conn *conn, struct net_buf *buf,
 	int err;
 
 	if (params) {
-		struct bt_att_req *req = params;
+		struct bt_att_req *req;
+
+		/* Allocate new request */
+		req = bt_att_req_alloc(K_FOREVER);
+		if (!req) {
```
[the remark this answers is not in the document]
Comment (Vudentz) [reply subsys/bluetooth/host/gatt.c]: It should probably be using a 30 seconds timeout instead like other parts of the code.

File: subsys/bluetooth/host/att.c (hunk 6 above)
[the remark this answers is not in the document]
Comment (Vudentz) [reply subsys/bluetooth/host/att.c]: Done, i end up moving this to use k_mem_slab as well.

**Review by joerchan — APPROVED** [review]

**Review by jhedberg — APPROVED** [review]
